services/lab-api/lab-api/api.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2018 Telstra Open Source
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+from flask import Flask, Response, request, jsonify
+from docker import DockerClient
+import requests
+import logging
+from urllib.parse import urlparse
+import os
+
+app = Flask(__name__)
+
+logger = logging.getLogger()
+
+FL_CONTAINER_NAME = "floodlight"
+API_CONTAINER_NAME = "lab-api"
+LAB_SERVICE = 'kilda_lab-service'
+docker = DockerClient(base_url='unix://var/run/docker.sock')
+
+HW_LOCKKEEPER_REST_HOST = os.environ.get("HW_LOCKKEEPER_REST_HOST")
+
+HW_LAB_ID = 0
+count = 0
+labs = {}
+
+
+def get_network_name():
+    cnt = docker.containers.get(API_CONTAINER_NAME)
+    return list(cnt.attrs['NetworkSettings']['Networks'].keys())[0]
+
+
+NETWORK_NAME = get_network_name()
+
+
+def make_container_name(id):
+    return 'lab_service-' + str(id)
+
+
+class Lab:
+    def __init__(self, lab_id, lab_def):
+        self.lab_def = lab_def
+        self.lab_id = lab_id
+
+        self.tgens = {}
+        for tgen_def in lab_def.get('active_traff_gens', []):
+            self.tgens[tgen_def['name']] = urlparse(tgen_def['control_endpoint'])
+
+        if lab_id != HW_LAB_ID:
+            name = make_container_name(lab_id)
+            env = {'LAB_ID': lab_id}
+            docker.containers.run(LAB_SERVICE, environment=env, name=name, privileged=True, detach=True)
+            docker.networks.get(NETWORK_NAME).connect(name, aliases=[name])
+
+    def destroy(self):
+        if self.lab_id != HW_LAB_ID:
+            cnt = docker.containers.get(make_container_name(self.lab_id))
+            cnt.stop()
+            cnt.remove()
+
+
+@app.route('/api/', methods=['GET'])
+def get_defined_labs():
+    return jsonify([lab_id for lab_id in labs.keys()])
+
+
+@app.route('/api', methods=['POST'])
+def create_lab():
+    global count
+
+    try:
+        is_hw = request.headers.get('Hw-Env', False)
+        if is_hw:
+            if str(HW_LAB_ID) in labs:
+                return Response('Hardware lab already defined', status=409)
+            else:
+                labs[str(HW_LAB_ID)] = Lab(HW_LAB_ID, request.get_json())
+                return jsonify({'lab_id': HW_LAB_ID})
+
+        count = count + 1
+        labs[str(count)] = Lab(count, request.get_json())
```

Comment:
...
use list for labs instead of dict
```
dummy = object()
labs.append(dummy)
index = labs.index(dummy)

labs[index] = Lab(index, reqiest.get_json())
```

On lab removal do `labs[index] = None`... no global variables and thread safe. If avoid passing id/index into Lab constructor - this schema will become much simpler(no need in dummy object).